tools/lsp/preview.rs

```diff
@@ -83,6 +88,140 @@ fn set_design_mode(enable: bool) {
     configure_design_mode(enable);
 }
 
+fn follow_child(element: &ElementRc) -> ElementRc {
+    let elem = element.borrow();
+    if elem.repeated.is_some() {
+        if let i_slint_compiler::langtype::ElementType::Component(base) = &elem.base_type {
+            return base.root_element.clone();
+        }
+    }
+    element.clone()
+}
+
+fn lsp_element_position(element: &ElementRc) -> (String, usize, usize, usize, usize) {
+    let e = &element.borrow();
+    e.node
+        .as_ref()
+        .and_then(|n| {
+            n.parent()
+                .filter(|p| p.kind() == i_slint_compiler::parser::SyntaxKind::SubElement)
+                .map_or_else(
+                    || Some(n.source_file.text_size_to_file_line_column(n.text_range().start())),
+                    |p| Some(p.source_file.text_size_to_file_line_column(p.text_range().start())),
+                )
+        })
+        .unwrap_or_else(|| (String::new(), 0, 0, 0, 0))
+}
+
+// triggered from the UI, running in UI thread
+pub fn select_element_at_impl(
+    x: f32,
+    y: f32,
+    component_instance: &ComponentInstance,
+    root_element: &ElementRc,
+) -> Option<ElementRc> {
+    let click_position = LogicalPoint::from_lengths(LogicalLength::new(x), LogicalLength::new(y));
+
+    for c in &root_element.borrow().children {
+        let c = follow_child(&c);
+
+        let Some(position) = component_instance.element_position(&c) else {
+            continue;
+        };
+        if position.contains(click_position) {
+            let (path, offset) = element_offset(&c);
+            let secondary_positions = component_instance.component_positions(path, offset);
+
+            set_selected_element(Some((&c, position)), secondary_positions);
+            let document_position = lsp_element_position(&c);
+            if !document_position.0.is_empty() {
+                ask_editor_to_show_document(
+                    document_position.0,
+                    document_position.1 as u32,
+                    document_position.2 as u32,
+                    document_position.3 as u32,
+                    document_position.4 as u32,
+                );
+            }
+            return Some(c.clone());
+        }
+    }
+
+    None
+}
+
+fn element_offset(element: &ElementRc) -> (PathBuf, u32) {
+    let Some(node) = &element.borrow().node else {
+        return (PathBuf::default(), 0);
```

Comment:
Should the function not return None?  if there is source code attached to it, we shouldn't redirect the user to it.